### PR TITLE
change docs to render stories out of context

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #13145 #13277 #10817 #8601 #12726 #14447 #14333 #13074 

## What I did
I've overhauled the way we render stories inline.

I've removed the need for prepareForInline:
When docs encounters a story it needs to render inline, it creates div element with a unique id on it.
Then it emits an event triggering the framework's renderer to render a story unto the targetted div.

## How to test

I'm likely missing some features, this is a POC / draft after all!

- does args and controls still work
- viewports
- toolbar addons
- HMR
- switching back and forth between canvas & docs
- does it work when docsOnly
- performance

Known issues:
- I'm not destroying / unmounting inline stories anymore.
  I do not know how bad this is, I assume memory leaks
- Preact seems to have issues with actions addon, might be pre-existing
- Vue2 doesn't re-render stories in docs-mode when controls / args changes
- Angular doesn't work at all

### CODE REVIEW

**Comment by ndelangen** [comment]
This does require a change in all framework's renderers. so the ones we just extracted out (@shilman) might need a patch, if we decide that's worth it.

**Comment by shilman** [comment]
🤯 this is awesome @ndelangen 🤯

**Comment by ndelangen** [comment]
@tooppaaa @gaetanmaisse @ThibaudAV 
Are you able to assist with making the required changes to angular's code in render?

**Comment by ThibaudAV** [comment]
When there is a navigation between two stroy of the same "stories" on the docs page all stories is loaded twice. I think it's a bug.  You can see it with the logs. Do you think it is related to these changes?

File: lib/core-client/src/preview/StoryRenderer.tsx [hunk 1]
```diff
@@ -77,13 +79,74 @@ export class StoryRenderer {
...
+    if (getDecorated) {
+      try {
+        const { applyLoaders, unboundStoryFn } = context;
+        console.time(`applyLoaders_${id}`);
+        const storyContext = await applyLoaders();
+        console.timeEnd(`applyLoaders_${id}`);
+        const storyFn = () => unboundStoryFn(storyContext);
+
+        if (targetDOMNode.querySelector('[data-is-loadering-indicator="true"')) {
+          targetDOMNode.querySelector('[data-is-loadering-indicator="true"').remove();
```
Comment (ThibaudAV) [inline lib/core-client/src/preview/StoryRenderer.tsx:127]: for angular it forces the full rendering every time. 🤔 
which differs from the canvas rendering which is optimized when only the args change

Comment (ThibaudAV) [inline lib/core-client/src/preview/StoryRenderer.tsx:127]: for angular it forces the full rendering every time. 🤔
which differs from the canvas rendering which is optimized when only the args change

**Comment by ndelangen** [comment]
@ThibaudAV What can we do with this PR versus this PR: https://github.com/storybookjs/storybook/pull/15002 and what has recently been happening on next?

**Comment by ThibaudAV** [comment]
if this is still intended to be taken for the next version. once rebase I would do the same for the one with angular (which will be greatly reduced ^^) and I would test it still works properly
what do you think?

**Comment by ndelangen** [comment]
When we have a solution that works across the board we can assess if this is truly a breaking change or not. If it is, it's a 7.0 feature.

We're not targeting 6.3. rather 6.4 or 7.0.
I'm not sure there will be a 6.4.

**Comment by shilman** [comment]
There will be a 6.4

**Comment by ThibaudAV** [comment]
as you like ;) I'm pretty confident that it will work for angular with this PR

**Comment by amartinez62** [comment]
Double check for feature flags if they aren't there then we need to add them and it should be off by default. Double check this works with Angular.

**Comment by shilman** [comment]
I genericized feature flagging here: https://github.com/storybookjs/storybook/pull/15375

**Comment by agrohs** [comment]
What is the latest on this PR?

**Comment by ndelangen** [comment]
I'm pretty sure this got merged via another PR somewhere in @tmeasday's work, can we close this @shilman ?

**Comment by shilman** [comment]
Thanks @ndelangen, closing this!

**Pull request closed without merging**